nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/api/AbstractSession.java

```diff
@@ -2031,6 +2036,94 @@ protected Object getDataModelField(DocumentRef docRef, String schema, String fie
         return null;
     }
 
+    @Override
+    public void makeRecord(DocumentRef docRef) {
+        Document doc = resolveReference(docRef);
+        if (isRecord(docRef)) {
+            // already a record, don't do anything
+            return;
+        }
+        checkPermission(doc, MAKE_RECORD);
+        DocumentModel docModel = readModel(doc);
+        notifyEvent(DocumentEventTypes.BEFORE_MAKE_RECORD, docModel, null, null, null, true, false);
+        doc.makeRecord();
+        docModel = readModel(doc);
+        notifyEvent(DocumentEventTypes.AFTER_MAKE_RECORD, docModel, null, null, null, true, false);
+    }
+
+    @Override
+    public boolean isRecord(DocumentRef docRef) {
+        Document doc = resolveReference(docRef);
+        checkPermission(doc, READ);
+        return doc.isRecord();
+    }
+
+    @Override
+    public void setRetainUntil(DocumentRef docRef, Calendar retainUntil) throws PropertyException {
+        Document doc = resolveReference(docRef);
+        if (!isRecord(docRef)) {
+            throw new PropertyException("Document is not a record");
+        }
+        Calendar current = doc.getRetainUntil();
+        if (Objects.equals(current, retainUntil)) {
+            // unchanged, don't do anything
+            return;
+        }
+        checkPermission(doc, SET_RETENTION);
+        Map<String, Serializable> options = new HashMap<>();
+        options.put(CoreEventConstants.RETAIN_UNTIL, retainUntil);
+        if (retainUntil != null) {
+            options.put("comment", retainUntil.toInstant().toString()); // for audit
+        }
+        DocumentModel docModel = readModel(doc);
+        notifyEvent(DocumentEventTypes.BEFORE_SET_RETENTION, docModel, options, null, null, true, false);
+        doc.setRetainUntil(retainUntil);
+        docModel = readModel(doc);
+        notifyEvent(DocumentEventTypes.AFTER_SET_RETENTION, docModel, options, null, null, true, false);
+    }
+
+    @Override
+    public Calendar getRetainUntil(DocumentRef docRef) {
+        Document doc = resolveReference(docRef);
+        checkPermission(doc, READ);
+        return doc.getRetainUntil();
+    }
+
+    @Override
+    public void setLegalHold(DocumentRef docRef, boolean hold) {
```

Comment:
This method must fire an event that will trigger an `elasticsearch` reindex else we end up with a desync index on `ecm:hasLegalHold`